File: crm/src/main/kotlin/it/polito/crm/entities/Professional.kt
package it.polito.crm.entities

import it.polito.crm.dtos.ProfessionalDTO
import it.polito.crm.utils.EmploymentState
import jakarta.persistence.*

@Entity
@Table(name = "professionals")
class Professional {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "professional_id", nullable = false)
    val professionalId: Long = 0

    @ElementCollection
    var skills = mutableSetOf<String>()

    @Column(name = "employment_state")
    var employmentState = EmploymentState.Unemployed

    @Column(name = "daily_rate")
    var dailyRate: Float = 0.0f

    @Column(name = "location")
    lateinit var location: String

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "contact_id", referencedColumnName = "contact_id")
    lateinit var contact: Contact

    @OneToMany(mappedBy = "professional", cascade = [CascadeType.ALL], orphanRemoval = true)
    val jobApplications = mutableSetOf<Application>()

    fun addJobApplication(jobOfferHistory: JobOfferHistory, status: ApplicationStatus = ApplicationStatus.Pending) {
        val application = Application()

        application.professional = this
        application.jobOfferHistory = jobOfferHistory
        application.status = status

        jobApplications.add(application)
        jobOfferHistory.candidates.add(application)
    }

    fun removeJobApplication(application: Application) {
        jobApplications.remove(application)
        application.jobOfferHistory.candidates.remove(application)
    }

    fun toDto(): ProfessionalDTO = ProfessionalDTO(
        this.professionalId,
        this.skills,
        this.employmentState,
        this.dailyRate,
        this.location,
        this.contact.toDto(),
        this.jobApplications.map { it.toDto() }.toMutableSet()
    )
}

File: crm_ui_client/src/api/crm/dto/Professional.ts
import {Contact, ContactRawData} from "./Contact";
import {Application, ApplicationRawData} from "./Application";

export enum EmploymentState {
    Employed, Unemployed, NotAvailable
}

interface ProfessionalRawData {
    professionalId: number | null
    skills: string[]
    employmentState: EmploymentState | null
    dailyRate: number
    location: string
    contact: ContactRawData | null
    jobApplications: ApplicationRawData[]
}

export class Professional implements ProfessionalRawData {
    professionalId: number | null
    skills: string[]
    employmentState: EmploymentState | null
    dailyRate: number
    location: string
    contact: ContactRawData | null
    jobApplications: Application[]

    constructor(
        professionalId: number | null,
        skills: string[],
        employmentState: EmploymentState | null,
        dailyRate: number, location: string,
        contact: ContactRawData | null,
        jobApplications: Application[]
    ) {
        this.professionalId = professionalId
        this.skills = skills
        this.employmentState = employmentState
        this.dailyRate = dailyRate
        this.location = location
        this.contact = contact
        this.jobApplications = jobApplications
    }

    static fromJsonObject(obj: ProfessionalRawData): Professional | null {
        try {
            const tmp: Array<Application> = obj.jobApplications.map((e: ApplicationRawData) => Application.fromJsonObject(e)!)
            const candidates: Application[] | null = []

            tmp.forEach((e: Application) => {
                candidates.push(e)
            })

            if (obj.contact ?? false) {
                console.log(obj.contact)
                return new Professional(
                    obj.professionalId,
                    obj.skills,
                    obj.employmentState,
                    obj.dailyRate,
                    obj.location,
                    Contact.fromJsonObject(obj.contact!),
                    candidates
                )
            } else {
                return new Professional(
                    obj.professionalId,
                    obj.skills,
                    obj.employmentState,
                    obj.dailyRate,
                    obj.location,
                    null,
                    candidates
                )
            }
        } catch (e) {
            console.error(e)
            return null
        }
    }
}
